cirq-core/cirq/ops/conditional_operation.py

```diff
@@ -0,0 +1,161 @@
+
+    def _resolve_parameters_(
+        self, resolver: 'cirq.ParamResolver', recursive: bool
+    ) -> 'ConditionalOperation':
+        new_sub_op = protocols.resolve_parameters(self._sub_operation, resolver, recursive)
+        return ConditionalOperation(new_sub_op, self._controls)
+
+    def __pow__(self, exponent: Any) -> 'ConditionalOperation':
+        new_sub_op = protocols.pow(self._sub_operation, exponent, NotImplemented)
+        if new_sub_op is NotImplemented:
+            return NotImplemented  # coverage: ignore
+        return ConditionalOperation(new_sub_op, self._controls)
+
+    def _circuit_diagram_info_(
+        self, args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> Optional['protocols.CircuitDiagramInfo']:
+        sub_args = protocols.CircuitDiagramInfoArgs(
+            known_qubit_count=args.known_qubit_count,
+            known_qubits=args.known_qubits,
+            use_unicode_characters=args.use_unicode_characters,
+            precision=args.precision,
+            qubit_map=args.qubit_map,
+        )
+        sub_info = protocols.circuit_diagram_info(self._sub_operation, sub_args, None)
+        if sub_info is None:
+            return NotImplemented  # coverage: ignore
+
+        wire_symbols = sub_info.wire_symbols + ('^',) * len(self._controls)
+        exponent_qubit_index = None
+        if sub_info.exponent_qubit_index is not None:
+            exponent_qubit_index = sub_info.exponent_qubit_index + len(self._controls)
+        elif sub_info.exponent is not None:
+            exponent_qubit_index = len(self._controls)
+        return protocols.CircuitDiagramInfo(
+            wire_symbols=wire_symbols,
+            exponent=sub_info.exponent,
+            exponent_qubit_index=exponent_qubit_index,
+        )
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            'cirq_type': self.__class__.__name__,
+            'controls': self._controls,
+            'sub_operation': self._sub_operation,
+        }
+
+    def _act_on_(self, args: 'cirq.ActOnArgs') -> bool:
+        def not_zero(measurement):
+            return any(i != 0 for i in measurement)
+
+        measurements = [args.log_of_measurement_results[str(key)] for key in self._controls]
+        if all(not_zero(measurement) for measurement in measurements):
+            protocols.act_on(self._sub_operation, args)
+        return True
+
+    def _with_key_path_(self, path: Tuple[str, ...]) -> 'ConditionalOperation':
+        return ConditionalOperation(
+            self._sub_operation, [protocols.with_key_path(k, path) for k in self._controls]
+        )
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]) -> 'ConditionalOperation':
+        return ConditionalOperation(
+            self._sub_operation,
+            [protocols.with_measurement_key_mapping(k, key_map) for k in self._controls],
+        )
+
+    def _control_keys_(self) -> Tuple[value.MeasurementKey, ...]:
```

Comment:
In parity with _measurement_keys_, should add control keys from the suboperation here too. Also note this protocol needs added to TaggedOperation. Aside, should add logic to ControlledOperation constructor to fail if it wraps a ConditionalOp or measurement.